libc/src/__support/GPU/allocator.cpp

```diff
@@ -27,21 +51,445 @@ void *rpc_allocate(uint64_t size) {
+    cpp::atomic_thread_fence(cpp::MemoryOrder::RELEASE);
+    if (gpu::get_lane_id() == uint32_t(cpp::countr_zero(mask)) &&
+        ref.release(cpp::popcount(mask))) {
+      T *p = ptr.load(cpp::MemoryOrder::RELAXED);
+      p->~T();
+      impl::rpc_free(p);
+      cpp::atomic_thread_fence(cpp::MemoryOrder::RELEASE);
+      ptr.store(nullptr, cpp::MemoryOrder::RELAXED);
+    }
+    gpu::sync_lane(lane_mask);
+  }
+
+  // Get the current value of the reference counter.
+  uint64_t use_count() { return ref.read(); }
+};
+
+// The global array used to search for a valid slab to allocate from.
+static GuardPtr<Slab> slots[ARRAY_SIZE] = {};
+
+// Tries to find a slab in the table that can support the given chunk size.
+static Slab *find_slab(uint32_t chunk_size) {
+  // We start at a hashed value to spread out different chunk sizes.
+  uint32_t start = impl::hash(chunk_size);
+  for (uint32_t offset = 0; offset < ARRAY_SIZE;) {
+    uint32_t index = (offset + start) % ARRAY_SIZE;
+
+    // If this slot is too full we exit early.
+    if (slots[index].use_count() >= Slab::available_chunks(chunk_size)) {
+      offset++;
+      sleep_briefly();
+      continue;
+    }
+
+    uint64_t lane_mask = gpu::get_lane_mask();
+    uint64_t uniform = gpu::match_any(lane_mask, index);
+    uint64_t reserved = 0;
+    Slab *slab =
+        slots[index].try_lock(lane_mask, uniform, reserved, chunk_size, index);
+    gpu::sync_lane(lane_mask);
+
+    // We successfully obtained a slab with enough space for our allocation.
+    // This guarantees that a call to Slab::allocate will always succeed.
+    if (slab && reserved <= Slab::available_chunks(chunk_size) &&
+        slab->get_chunk_size() == chunk_size)
+      return slab;
+
+    // We encountered either a full slab or an slab with an incompatible chunk
+    // size. Move to the next slot.
+    if (slab && reserved > Slab::available_chunks(chunk_size) &&
+        slab->get_chunk_size() == chunk_size) {
+      slots[index].unlock(gpu::get_lane_mask(), gpu::get_lane_mask() & uniform);
+      offset++;
+    }
+
+    // Malloc returned a null pointer and we are out-of-memory.
+    if (!slab && reserved == cpp::numeric_limits<uint64_t>::max())
+      return nullptr;
+
+    // The slab is in the process of being initialized. Start at the beginning
+    // to prevent too many slab allocations from happening at once.
+    if (!slab && reserved == 0)
+      offset = 0;
+    sleep_briefly();
+  }
+  return nullptr;
+}
+
+// Release the lock associated with a given slab.
+static void release_slab(Slab *slab) {
+  uint32_t index = slab->get_global_index();
+  uint64_t lane_mask = gpu::get_lane_mask();
+  uint64_t uniform = gpu::match_any(lane_mask, index);
+  slots[index].unlock(lane_mask, uniform);
+}
 
 namespace gpu {
 
-void *allocate(uint64_t size) { return rpc_allocate(size); }
+void *allocate(uint64_t size) {
+  if (!size)
+    return nullptr;
+
+  // Allocations larger than a single slab go directly to memory.
```

Comment:
The wording here doesn't quite match SLAB_SIZE / 2 but I'm struggling to come up with something better.

Allocations that would consume a full slab or more?
Allocations that are too large to be 'slabbed'?